llms.md

```diff
@@ -0,0 +1,1039 @@
+        ctx.db.user().identity().update(User { name: Some(name), ..user });
+        Ok(())
+    } else {
+        Err("Cannot set name for unknown user".to_string())
+    }
+}
+
+// Reducer to send a message
+#[reducer]
+pub fn send_message(ctx: &ReducerContext, text: String) -> Result<(), String> {
+    let text = validate_message(text)?; // Basic validation
+    log::info!("Received message: {}", text);
+    // Insert the new message into the table
+    ctx.db.message().insert(Message {
+        sender: ctx.sender,
+        text,
+        sent: ctx.timestamp,
+    });
+    Ok(())
+}
+
+// Reducer called when a client connects
+#[reducer(client_connected)]
+pub fn client_connected(ctx: &ReducerContext) {
+    if let Some(user) = ctx.db.user().identity().find(ctx.sender) {
+        // Mark existing user as online
+        ctx.db.user().identity().update(User { online: true, ..user });
+    } else {
+        // Insert a new user record for the new connection
+        ctx.db.user().insert(User {
+            name: None,
+            identity: ctx.sender,
+            online: true,
+        });
+    }
+}
+
+// Reducer called when a client disconnects
+#[reducer(client_disconnected)]
+pub fn client_disconnected(ctx: &ReducerContext) {
+    if let Some(user) = ctx.db.user().identity().find(ctx.sender) {
+        // Mark user as offline
+        ctx.db.user().identity().update(User { online: false, ..user });
+    } else {
+        log::warn!("Disconnect event for unknown user: {:?}", ctx.sender);
+    }
+}
+
+// Reducer called once when the module is loaded/database is created
+#[reducer(init)]
+pub fn initialize_database(ctx: &ReducerContext) {
+    log::info!("Database Initializing! Module Identity: {}", ctx.identity());
+    // Perform one-time setup, like inserting initial data if tables are empty
+    if ctx.db.user().count() == 0 {
+        // Add an admin user or default settings
+    }
+}
+
+// Helper validation functions (example)
+fn validate_name(name: String) -> Result<String, String> {
+    if name.is_empty() { Err("Name cannot be empty".to_string()) } else { Ok(name) }
+}
+
+fn validate_message(text: String) -> Result<String, String> {
+    if text.is_empty() { Err("Message cannot be empty".to_string()) } else { Ok(text) }
+}
+
+##### Error Handling: `Result` vs. Panic
+
+Reducers can fail by returning `Err` or by panicking. Both methods trigger a transaction rollback.
+*   **Returning `Err(E)**:** This is the preferred way to handle *expected* failures (e.g., invalid input, failed validation). The error message (`E` must implement `Display`) is propagated back to the calling client and can be observed in the `ReducerEventContext` status.
+*   **Panicking:** This typically represents an *unexpected* or unrecoverable error (e.g., assertion failure, bug). While it also rolls back the transaction, the client might not receive a specific error message, potentially just seeing a generic failure or disconnection.
```

Comment:
This is not accurate. The client will always see an error message derived from the argument to `panic!`, and will never be disconnected due to a panic. The only reason to prefer returning `Result::Err` over panicking (aside from aesthetics) is that a panic will destroy the WASM instance, so the next reducer call will have to construct a new WASM instance, and thus see increased latency.

Reply:
Ah yes, you are correct. Didn't catch this one, thanks for the flag.